### PR TITLE
Reflecting BC

### PR DESCRIPTION
**Pull request by BenWibking**
New version of https://github.com/parthenon-hpc-lab/athenapk/pull/58. (Renamed the branch and didn't realize the PR would be closed automatically.)

### CODE REVIEW

**Comment by BenWibking** [comment]
@pgrete @forrestglines I'd also like to get this merged in, since it's needed for my simulations. Let me know what would need to be polished to make it suitable for merging to main.

**Comment by forrestglines** [comment]
> @pgrete @forrestglines I'd also like to get this merged in, since it's needed for my simulations. Let me know what would need to be polished to make it suitable for merging to main.

So this is working and ready for review I'm guessing? I'll remove the WIP if that's the case.

File: src/bc.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                                : bounds.js(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x2)
+                                ? bounds.je(parthenon::IndexDomain::interior)
+                                : bounds.je(parthenon::IndexDomain::entire)},
+      parthenon::IndexRange{IsDomainBound(pmb, parthenon::BoundaryFace::inner_x3)
+                                ? bounds.ks(parthenon::IndexDomain::interior)
+                                : bounds.ks(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x3)
+                                ? bounds.ke(parthenon::IndexDomain::interior)
+                                : bounds.ke(parthenon::IndexDomain::entire)}};
```
Comment (forrestglines) [inline src/bc.hpp:52]: This snippet of code seems like it could belong in Parthenon. I'm not sure about the name, `GetZonesInDomain` or something else referencing the problem domain might be better.

File: src/bc.hpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  pmb->par_for_bndry(
+      label, nvar, domain, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        if (!q.IsAllocated(l)) return;
+        if (TYPE == BCType::Reflect) {
+          q(l, k, j, i) =
+              (is_normal ? -1.0 : 1.0) *
+              q(l, X3 ? offset - k : k, X2 ? offset - j : j, X1 ? offset - i : i);
+        } else {
+          q(l, k, j, i) = q(l, X3 ? ref : k, X2 ? ref : j, X1 ? ref : i);
```
Comment (forrestglines) [inline src/bc.hpp:111]: I might wonder about executing these ternary statements inside the kernel and wonder if the CUDA compiler is smart enough to compile them out.

However, this kernel is certainly memory bandwidth limited so the extra integer computes are irrelevant

File: src/bc.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#define BC_HPP_
+//========================================================================================
+// AthenaPK - a performance portable block structured AMR astrophysical MHD code.
+// Copyright (c) 2021, Athena-Parthenon Collaboration. All rights reserved.
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+//! \file bc.hpp
+//  \brief Custom boundary conditions for AthenaPK
+//
+// Computes reflecting boundary conditions using AthenaPK's cons variable pack.
```
Comment (forrestglines) [inline src/bc.hpp:11]: If this is only for reflecting boundary conditions, should this file be renamed? Should we have a `bc` folder?

Or maybe this file can keep this name as a place holder for other commonly used boundary functions in AthenaPK

File: src/bc.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                                ? bounds.ke(parthenon::IndexDomain::interior)
+                                : bounds.ke(parthenon::IndexDomain::entire)}};
+}
+
+enum class BCSide { Inner, Outer };
+enum class BCType { Outflow, Reflect };
+
+template <parthenon::CoordinateDirection DIR, BCSide SIDE, BCType TYPE>
+void ApplyBC(parthenon::MeshBlock *pmb, parthenon::VariablePack<Real> &q,
+             parthenon::IndexRange &nvar, const bool is_normal, const bool coarse) {
```
Comment (forrestglines) [inline src/bc.hpp:60]: Should we instead template on:
```suggestion
template <parthenon::BoundaryFace FACE, BCType TYPE>
void ApplyBC(parthenon::MeshBlock *pmb, parthenon::VariablePack<Real> &q,
             parthenon::IndexRange &nvar, const bool is_normal, const bool coarse) {
```
and remove the `BCSide` enum? Would this work with your other code?

Comment (forrestglines) [reply src/bc.hpp:60]: I'm not quite sure how this function is used/should be used. In this PR it looks like it's only used to apply reflecting boundary conditions. Don't all the other pgen's have outflow boundary conditions applied as the default? Unless I've misunderstood something.

**Review by forrestglines — CHANGES_REQUESTED** [review]
I think this PR needs a test to exercise reflecting boundary conditions. A shock reflecting off a boundary should be pretty easy to add with the other Sod shock tube tests. I can do that next week if need be.

We can also add quick documentation.

I've also commented on a few code snippets that I think could be placed in Parthenon but I don't think those should hold up this PR. We can add those snippets to Parthenon later.

**Comment by forrestglines** [comment]
@BenWibking Is there a branch/pgen using the reflecting boundary conditions? If it's the precipitator boxes, could you link that here?

**Comment by BenWibking** [comment]
> @BenWibking Is there a branch/pgen using the reflecting boundary conditions? If it's the precipitator boxes, could you link that here?

https://github.com/parthenon-hpc-lab/athenapk/pull/66

**Comment by pgrete** [comment]
User boundary conditions recently got an (internal) overhaul in Parthenon, which might make interfacing cleaner (and reduce the amount of code required on our end).
I'm willing to take a stab at it (and adding a a double Mach reflection pgen for testing) if that sounds good.

**Comment by BenWibking** [comment]
> User boundary conditions recently got an (internal) overhaul in Parthenon, which might make interfacing cleaner (and reduce the amount of code required on our end). I'm willing to take a stab at it (and adding a a double Mach reflection pgen for testing) if that sounds good.

Sure, if you want to create a new PR with that, I can close this one.
